threedi_api_client/files.py

```diff
@@ -151,3 +153,112 @@ def download_fileobj(
+    url: str,
+    file_path: Path,
+    timeout: float = 5.0,
+    pool: Optional[urllib3.PoolManager] = None,
+) -> int:
+    """Upload a file at specified file path to a url.
+
+    The upload is accompanied by an MD5 hash so that the file server can check
+    the integrity of the file. This function is not inteded for files larger than
+    1 GB.
+
+    Args:
+        url: The url to upload to.
+        file_path: The file path to read data from.
+        timeout: The total timeout in seconds.
+        pool: If not supplied, a default connection pool will be
+            created with a retry policy of 3 retries after 1, 2, 4 seconds.
+
+    Returns:
+        The total amount of uploaded bytes.
+
+    Raises:
+        IOError: Raised if the provided file is incompatible or empty.
+        openapi_client.exceptions.ApiException: raised on unexpected server
+            responses (HTTP status codes other than 206, 413, 429, 503)
+        urllib3.exceptions.HTTPError: various low-level HTTP errors that persist
+            after retrying: connection errors, timeouts, decode errors,
+            invalid HTTP headers, payload too large (HTTP 413), too many
+            requests (HTTP 429), service unavailable (HTTP 503)
+    """
+    # cast string to Path if necessary
+    if isinstance(file_path, str):
+        file_path = Path(file_path)
+
+    # open the file
+    with file_path.open("rb") as fileobj:
+        size = upload_fileobj(url, fileobj, timeout=timeout, pool=pool)
+
+    return size
+
+
+def upload_fileobj(
+    url: str,
+    fileobj: BinaryIO,
+    timeout: float = 5.0,
+    pool: Optional[urllib3.PoolManager] = None,
+) -> int:
+    """Upload a file object to a url.
+
+    The upload is accompanied by an MD5 hash so that the file server can check
+    the integrity of the file. This function is not inteded for files larger than
+    1 GB.
+
+    Args:
+        url: The url to upload to.
+        fileobj: The (binary) file object to read from.
+        size: The size of the file, in bytes.
+        timeout: The total timeout in seconds.
+        pool: If not supplied, a default connection pool will be
+            created with a retry policy of 3 retries after 1, 2, 4 seconds.
+
+    Returns:
+        The total amount of uploaded bytes.
+
+    Raises:
+        IOError: Raised if the provided file is incompatible or empty.
+        openapi_client.exceptions.ApiException: raised on unexpected server
+            responses (HTTP status codes other than 206, 413, 429, 503)
+        urllib3.exceptions.HTTPError: various low-level HTTP errors that persist
+            after retrying: connection errors, timeouts, decode errors,
+            invalid HTTP headers, payload too large (HTTP 413), too many
+            requests (HTTP 429), service unavailable (HTTP 503)
+    """
+    # There are two ways to upload in S3 (Minio):
+    # - PutObject: put the whole object in one time
+    # - multipart upload: requires presigned urls for every part
```

Comment:
I've just looked into multipart for Minio, but could not find it for presigned URL's in the Python (client) lib. If presigned-url's can be used we can alter the endpoint for retrieving them to respond with multiple upload presigned URL's.

Reply:
Update: https://github.com/minio/minio-py/blob/fe78f929454f536942dc13169b20c3c30b8e65ed/minio/api.py#L1603

Reply:
I could find this blog specifically to MinIO:

https://vsoch.github.io/2020/s3-minio-multipart-presigned-upload/

> However, minio-py doesn’t support generating anything for pre-signed multipart, so in this case we need to interact with s3 via boto3.

Indeed is seems that we can return a list of presigned urls, but then you need to know 1) the number of each part and 2) the size (in bytes) of each part before creating the signed urls. Also the client is then responsible for doing the additional call "completeMultipartUpload".

Something like this could be done:

```
endpoint/<pk>/upload-gridadmin/
endpoint/<pk>/upload-gridadmin/initiate-multipart/
endpoint/<pk>/upload-gridadmin/upload-part/?id=42&part_number=5&size=512341
endpoint/<pk>/upload-gridadmin/complete-multipart/?id=42
endpoint/<pk>/upload-gridadmin/abort-multipart/?id=42
``

Each returning a signed url.